### PR TITLE
Allow for introspection into arguments

### LINKED ISSUE #71: Allow for introspection into arguments
e.g. say we only want to verify one element in a dictionary:

	def my_func(argument_to_verify):
	    return argument_to_verify.get('some_key') == 'Bob Barker'
	allow(foo).bar.with_args_that_match(my_func)
	foo.bar({'some_key': 'Bob Barker', 'some_other_key': 'Drew Carey'})

Open to feedback on the naming.

### PR DESCRIPTION
**Pull request by toddsifleet**
* Fixes #71
* Allow users to define custom argument matchers

### CODE REVIEW

File: doubles/allowance.py [hunk 1]
```diff
@@ -119,6 +120,17 @@ def with_args(self, *args, **kwargs):
         self.verify_arguments()
         return self
 
+    def with_args_that_match(self, matching_function):
+        """
+        Define a custom function for testing arguments
+
+        :param func matching_function:  The funciton used to test arguments passed to the stub.
```
Comment (charlax) [inline doubles/allowance.py]: typo

File: doubles/allowance.py [hunk 2]
```diff
@@ -119,6 +120,17 @@ def with_args(self, *args, **kwargs):
         self.verify_arguments()
         return self
 
+    def with_args_that_match(self, matching_function):
+        """
+        Define a custom function for testing arguments
```
Comment (charlax) [inline doubles/allowance.py]: should be on the previous line, see pep257

File: doubles/allowance.py [hunk 3]
```diff
@@ -119,6 +120,17 @@ def with_args(self, *args, **kwargs):
         self.verify_arguments()
         return self
 
+    def with_args_that_match(self, matching_function):
```
Comment (charlax) [inline doubles/allowance.py]: `with_args_callback`? Not sure if that's better, but match seems to imply something like a regex. I'll try to think about it.

File: doubles/allowance.py [hunk 4]
```diff
@@ -179,6 +193,20 @@ def satisfy_exact_match(self, args, kwargs):
 
         return True
 
+    def satisfy_custom_matcher(self, args, kwargs):
+        """
+        Returns a boolean indicating whether or not the stub will accept the provided arguments.
```
Comment (charlax) [inline doubles/allowance.py]: `Return`, and on the previous line. I know the rest of the codebase follow this other convention, but probably safer to use the more widely accepted pep257 one.

File: doubles/allowance.py (hunk 3 above)
Comment (toddsifleet) [inline doubles/allowance.py]: I agree about the name not being great.  I'm not a huge fan of `with_args_callback` either.  I will think about it more too.

Comment (toddsifleet) [inline doubles/allowance.py]: @charlax maybe `with_args_that_satisfy`?

Comment (chucknthem) [inline doubles/allowance.py]: with_args_validator?

**Comment by toddsifleet** [comment]
I squashed merged this onto master.

**Pull request closed without merging**